File: test/revwalk_test.erl
-module(revwalk_test).
-compile([export_all]).
-include_lib("eunit/include/eunit.hrl").

repo_test_() ->
    case os:getenv("GEEF_RESOURCES") of
	false ->
	    [];
	_ ->
	    {foreach, fun start/0, fun stop/1, [fun amount_test/1]}
    end.

start() ->
    BasePath = os:getenv("GEEF_RESOURCES"),
    Path = filename:join([BasePath, "testrepo.git"]),
    {ok, Repo} = geef_repo:open(Path),
    Repo.

count_walk(Walk, Acc) ->
    case geef_revwalk:next(Walk) of
	{ok, _} ->
	    count_walk(Walk, Acc + 1);
	{error, iterover} ->
	    Acc
    end.

amount_test(Repo) ->
    {ok, Walk} = geef_repo:revwalk(Repo),
    Id = geef_oid:parse("a4a7dce85cf63874e984719f4fdd239f5145052f"),
    ok = geef_revwalk:push(Walk, Id),
    Count = count_walk(Walk, 0),
    geef_revwalk:stop(Walk),
    [?_assertEqual(6, Count)].

stop(Repo) ->
    geef_repo:stop(Repo).

File: test/revwalk_test.exs
defmodule RevwalkTest do
  use ExUnit.Case
  use Geef
  import RepoHelpers


  setup do
    {repo, path, head, boring_ancestor} = tmp_commit_line
    Process.link(repo)
    {:ok, walk} = Repository.revwalk(repo)
    on_exit(fn -> File.rm_rf!(path) end)
    {:ok, [walk: walk, head: head, boring_ancestor: boring_ancestor]}
  end

  test "walk and count", meta do
    walk = meta[:walk]
    Revwalk.push(walk, meta[:head])
    Revwalk.hide(walk, meta[:boring_ancestor])
    assert 3 = count_walk(walk)
  end

  defp count_walk walk, acc \\ 0
  defp count_walk walk, acc do
    case Revwalk.next(walk) do
      {:ok, _} -> count_walk(walk, acc + 1)
      {:error, :iterover} -> acc
    end
  end
end
